### PR TITLE
Throttle the image request queue while the map is moving to improve performance

### PR DESCRIPTION
**Pull request by AdamS108**
The image request queue can cause frame rate glitches while the map is moving interactively (panning, zooming, or rotating). To mitigate this, allow the map renderer to throttle the number of requests processed by the image request queue per frame until the map stops moving again.

The default throttling limit set by this PR is 8 image requests per frame, but it is configurable either by setting the value of `Config.MAX_PARALLEL_IMAGE_REQUESTS_PER_FRAME_WHILE_THROTTLED` or by using the `maxParallelImageRequestsPerFrameWhileThrottled` property.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to [related issues](https://github.com/maplibre/maplibre-gl-js/issues/96).
 - [x] Include before/after visuals or gifs if this PR includes visual changes.  
   Video: https://youtu.be/a51bc841ouk
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Post benchmark scores.  
   The most significant difference is that the MapIdle benchmark increased due to the tradeoff between frame rate and image load time. This PR increases frame rate while panning and zooming the map but it does so by throttling the image request queue, thereby increasing image load time (but only while panning and zooming).  
   benchmark result
 - [x] Manually test the debug page.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/util/ajax.ts [hunk 1]
```diff
@@ -351,13 +351,51 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
+// use a lambda function to determine when the queue should be throttled (e.g. when isMoving())
+// and manually calling processImageRequestQueue() in the render loop.
+
+export type ImageQueueThrottleControlCallback = () => boolean;
+
+let imageQueueThrottleControlCallbackHandleCounter: number = 0;
+const imageQueueThrottleControlCallbacks = [];
+
+// Install a callback to control when image queue throttling is desired (e.g. when the map view is moving).
+export const installImageQueueThrottleControlCallback = function (callback: ImageQueueThrottleControlCallback): number /*callbackHandle*/ {
```
Comment (HarelM) [inline src/util/ajax.ts:380]: I would expect to have a file dedicated for the queue if we see that there are more functionally added and avoid writing the logic in the ajax file...

Comment (AdamSzofranMSFT) [reply src/util/ajax.ts]: I moved all of the queue related code from ajax.ts into a new image_request_queue.ts file. It's still just a collection of loose functions, not a class, but I hope it addresses your main concern about reducing comingling of functionality.

Comment (AdamSzofranMSFT) [reply src/util/ajax.ts]: @HarelM (In case my previous comment doesn't ping you.)

**Comment by HarelM** [comment]
I think this PR might conflict with #1805, so once the other PR is merged let's focus on this one.

File: src/util/ajax.ts (hunk 1 above)
Comment (HarelM) [reply src/util/ajax.ts]: This file has all kind of variables laying around, any chance you can turn it into a class? at least the part where you have a state, I don't mind if the pure functions are kept as standalone, but please move the part where you have variables and state to a class.

Comment (AdamSzofranMSFT) [reply src/util/ajax.ts]: @HarelM, thanks for the feedback. I can collect the loose data into a class.

However, I've been having second thoughts about my earlier change to split the queue functionality into its own file. Unfortunately, that makes my changes harder to see since they're buried in the new file. Perhaps I should undo that file split, put everything back into ajax.ts, and do the loose data collection into a class. That way my diffs are more apparent (and smaller). Later, we can do another PR that only does the file split without polluting it with other changes. Thoughts?

Comment (HarelM) [reply src/util/ajax.ts]: If you think two PR are better, I'm totally fine with that. What ever you feel is better.

Comment (AdamSzofranMSFT) [reply src/util/ajax.ts]: @HarelM latest commits move the queue code back into ajax.ts but in the form of a class with pure wrapper functions that call into it.

File: src/util/ajax.ts [hunk 2]
```diff
@@ -324,81 +324,181 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
+    private throttleControlCallbacks = [];
+
+    public resetImageRequestQueue() {
+        this.imageQueue = [];
+        this.currentParallelImageRequests = 0;
+    }
+
+    // Install a callback to control when image queue throttling is desired (e.g. when the map view is moving).
+    public installThrottleControlCallback(callback: ImageQueueThrottleControlCallback): number /*callbackHandle*/ {
+        const handle = this.throttleControlCallbackHandleCounter++;
```
Comment (HarelM) [inline src/util/ajax.ts:357]: Can this reach a number too big to become negative?

File: src/util/ajax.ts [hunk 3]
```diff
@@ -324,81 +324,181 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
+        const index = this.throttleControlCallbacks.indexOf(callbackHandle, 0);
+        if (index >= 0) {
+            this.throttleControlCallbacks.splice(index, 1);
         }
-        requestParameters.headers.accept = 'image/webp,*/*';
     }
 
-    // limit concurrent image loads to help with raster sources performance on big screens
-    if (numImageRequests >= config.MAX_PARALLEL_IMAGE_REQUESTS) {
+    // Check to see if any of the installed callbacks are requesting the queue to be throttled.
```
Comment (HarelM) [inline src/util/ajax.ts:370]: Use jsdocs commets style.

File: src/util/ajax.ts [hunk 4]
```diff
@@ -324,81 +324,181 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
+        callback: GetImageCallback
+    ): Cancelable {
+        if (webpSupported.supported) {
+            if (!requestParameters.headers) {
+                requestParameters.headers = {};
+            }
+            requestParameters.headers.accept = 'image/webp,*/*';
+        }
+
+        const theQueue = this;
```
Comment (HarelM) [inline src/util/ajax.ts:386]: Not sure I underrated this assignment, can't we simply use `this` where needed?

**Comment by HarelM** [comment]
This looks good, thanks!
There's a test inside the side to check if there should be throttling. Is there a way that this world cause the queue to get struck and not continue the processing (at all, nit just to facilitate the throttling)?

File: src/util/ajax.ts (hunk 2 above)
Comment (AdamSzofranMSFT) [reply src/util/ajax.ts:357]: Sure, it can overflow but that shouldn't cause a problem. What matters is that each handle is unique among all currently installed callbacks. Most applications will only install zero or one callbacks so it would be a very, very unusual situation for enough callbacks to be installed to cause overflow and, even then, negative handle values wouldn't cause a problem.

What *COULD* cause a problem is if the handle value wraps around and starts colliding with previously installed (and still existing) callbacks. I suppose I could add additional logic to check for collisions, but I'll let you decide if that's worth the extra code or not. The probability seems vanishingly low to me, although I concede it is non-zero.

File: src/util/ajax.ts (hunk 4 above)
Comment (AdamSzofranMSFT) [reply src/util/ajax.ts:386]: Let me know if there's a better way to do this. I'm relatively new to JS/TS so I might be showing my ignorance.

File: src/util/ajax.ts (hunk 3 above)
Comment (AdamSzofranMSFT) [reply src/util/ajax.ts]: Added

**Comment by AdamSzofranMSFT** [comment]
> There's a test inside the side to check if there should be throttling. Is there a way that this world cause the queue to get struck and not continue the processing (at all, nit just to facilitate the throttling)?

The short answer is "yes, but..."

Here's the longer answer.

The previous implementation of the queue would always start processing additional items as soon as the previous items completed (up to MAX_PARALLEL_IMAGE_REQUESTS at a time). Unfortunately, in interactive scenarios with maps containing many raster layers, this could result in an unbounded number of raster requests being processed per frame. One symptom of this is laggy/glitchy frame rate while a user and panning and zooming the map.

The modified queue implementation in this PR allows the caller (in this case, the map) to schedule the queue processing as necessary to maintain smooth(er) interactive frame rates (rather than the queue processing running without bound). This only happens if the caller installs a callback that requests throttling to be enabled. The responsibility of someone installing that callback is that they are now responsible for calling `processingImageQueueRequests` on a periodic basis (i.e. once per frame).

So, to answer your question, yes, it's possible to request throttling then neglect to process the queue at all, resulting in a stalled queue. But that effectively violates the contract.

File: src/util/ajax.ts (hunk 4 above)
Comment (HarelM) [reply src/util/ajax.ts:409]: When using lambdas "this" is captured.
So doing something like `() => ...` would capture this inside.

**Comment by HarelM** [comment]
My question is if there's a situation where the if is executed so the queue will not prices the next item, and then due to changes in the condition (I.e isMoving will change) the queue will not process new requests further on.
Not sure I manage to pass my point here... :-/
Also, is this now enabled by default or there's a map option that needs to be set to enable this?as I'm not sure this is the requested behavior for everyone - if someone prefers the glitchy map as the new locally the map is moving to still be presented faster, right?

File: src/util/ajax.ts (hunk 4 above)
Comment (AdamSzofranMSFT) [reply src/util/ajax.ts:409]: There were two instances where I used that syntax. I fixed one instance. However, the one you called out above can't be fixed I don't think. It's because the reference to the queue occurs inside an object initialization where `this` actually refers to the object being initialized, not the queue.

```
        const theQueue = this;
        const queued = { 
            requestParameters,
            callback,
            cancelled: false,
            completed: false,
            theQueue,
            cancel() {
                this.cancelled = true;  // in this case, `this` refers to the object being initialized

                if (!theQueue.isThrottled()) {
                    theQueue.processQueue(config.MAX_PARALLEL_IMAGE_REQUESTS);
                }
            }
        };
```

**Comment by AdamSzofranMSFT** [comment]
> My question is if there's a situation where the if is executed so the queue will not prices the next item, and then due to changes in the condition (I.e isMoving will change) the queue will not process new requests further on. Not sure I manage to pass my point here... :-/

As currently implemented, there are only two cases:  
  1. If the map isn't moving then requests are allowed to be processed without bound, up to `MAX_PARALLEL_IMAGE_REQUESTS` at a time in parallel.  
  2. If the map is moving then only up to `MAX_PARALLEL_IMAGE_REQUESTS_PER_FRAME_WHILE_THROTTLED` request will be processed on the current frame, with the remaining ones being processed on subsequent frames.

If the map starts and stops moving then the behavior toggles between the two cases above. But I don't see a case where the queue will completely stall.

**Comment by AdamSzofranMSFT** [comment]
> Also, is this now enabled by default or there's a map option that needs to be set to enable this?as I'm not sure this is the requested behavior for everyone - if someone prefers the glitchy map as the new locally the map is moving to still be presented faster, right?

The behavior is enabled by default, yes. The new behavior is configurable via `MAX_PARALLEL_IMAGE_REQUESTS_PER_FRAME_WHILE_THROTTLED` but it can't currently be completely disabled. I suppose I could add a way to do that. But making the "off" behavior the default means that the new behavior wouldn't get much testing and might be accidentally broken in some subtle way that evades unit testing. (???)

File: src/util/ajax.ts (hunk 4 above)
Comment (AdamSzofranMSFT) [reply src/util/ajax.ts:409]: Well, my attempt to remove one of the instances of this pattern failed the unit tests. So... back to the drawing board.

**Comment by HarelM** [comment]
There's another PR now to replace ajax callback with promises :-(
@smellyshovel will you be able to handle the conflicts in case this gets merged?

**Comment by AdamSzofranMSFT** [comment]
@HarelM where does this PR stand? I tried to answer or address your questions and concerns. Are you satisfied or are there still outstanding issues? Thanks.

**Comment by smellyshovel** [comment]
@AdamSzofranMSFT, here's the mentioned PR: https://github.com/maplibre/maplibre-gl-js/pull/2038

@HarelM, I don't really understand the point of this PR, but i feel like it has something to do with the image request queue. From my point of view, in the existing implementation the image request queue does not work properly (does it work at all?). So my vision of the flow is the following:

1. Fix broken image queue (the original implementation) under the scope of my PR
2. Merge it
3. Add the functionality from this PR on top of it

Another strategy would be to skip the second step and make this PR a part of the refactoring one, but there are obvious cons to such a solution.

Long story short - you decide what's better based on importance/relevance of a given PR. I'll try to keep with any decision you make, but if you ask me, then be I in your shoes, I'd first merge the other one. But on the other hand, I really don't like being a blocker for other useful PRs with my bloated ones.

File: src/util/ajax.ts [hunk 5]
```diff
@@ -324,81 +324,249 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
         const queued = {
             requestParameters,
             callback,
             cancelled: false,
-            cancel() { this.cancelled = true; }
+            completed: false,
+            theQueue,
+            cancel() {
```
Comment (HarelM) [inline src/util/ajax.ts:416]: This is an object not a simple data type.
The fact that a queue object "knows" the queue is less than ideal and indicate of a problematic design.
I would consider creating an id for every queue item and having a cancel method on the queue with the id instead of here.

File: src/util/ajax.ts [hunk 6]
```diff
@@ -324,81 +324,249 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
 
-        if (err) {
-            callback(err);
-        } else if (data) {
-            const decoratedCallback = (imgErr?: Error | null, imgResult?: CanvasImageSource | null) => {
-                if (imgErr != null) {
-                    callback(imgErr);
-                } else if (imgResult != null) {
-                    callback(null, imgResult as (HTMLImageElement | ImageBitmap), {cacheControl, expires});
+        const cancelRequest = (request: any) => {
```
Comment (HarelM) [inline src/util/ajax.ts:473]: Request need a type, it shouldn't be "any"

**Comment by HarelM** [comment]
@AdamSzofranMSFT I would consider opening a PR against @smellyshovel's fork.
The changes in callback will affect this dramatically so I think it would be better to write this against the other PR instead of main.
This PR still has some fixing that need implemented and I think it will be easier to do it with promises instead of callbacks.
This will allow to move the queue outside ajax file and treat it differently.
Once both PRs are reviewed separately we can merge them in whatever order we see fit.

**Comment by AdamS108** [comment]
@HarelM, @zhangyiatmicrosoft will be taking over this PR. I believe his intent is to have me abandon this one and he'll submit a new one from one of his branches. Thank you. -Adam

**Comment by AdamS108** [comment]
@HarelM please see the [replacement PR](https://github.com/maplibre/maplibre-gl-js/pull/2097) submitted by @zhangyiatmicrosoft. I'm now abandoning this one.

**Comment by HarelM** [comment]
Thanks for the update!
